src/loam/allocation.rs

```diff
@@ -346,174 +516,164 @@ ascent! {
-        let c2_4__4_6 = allocator().hash4(Tag::Cons.value(), c2_4, Tag::Cons.value(), c4_6);
-
-        let mut test = |input, expected_output, cons_count| {
-            let mut prog = AllocationProgram::default();
-
-            prog.input_expr = vec![(WidePtr(Tag::Cons.value(), input),)];
-            prog.run();
-
-            println!("{}", prog.relation_sizes_summary());
-
-            assert_eq!(
-                vec![(WidePtr(Tag::Cons.value(), expected_output),)],
-                prog.output_expr
-            );
-            assert_eq!(cons_count, prog.cons_mem.len());
-            assert_eq!(cons_count, prog.cons_digest_mem.len());
-
-            // This will often fail in the first pass. We need to fix up the memory and provide it as a hint to the
-            // second pass.
-            //assert!(prog.cons_mem_is_contiguous());
-
-            prog
+    fn wide_ptr(tag: LE, digest: [LE; 8]) -> WidePtr {
+        WidePtr(Wide::widen(tag), Wide(digest))
+    }
+
+    fn read_wideptr(zstore: &mut ZStore<BabyBear, LurkChip>, src: &str) -> WidePtr {
+        let ZPtr { tag, digest } = zstore.read(src).unwrap();
+        wide_ptr(tag.elt(), digest)
+    }
+
+    fn test_aux0(
+        zstore: &ZStore<BabyBear, LurkChip>,
+        input: WidePtr,
+        expected_output: WidePtr,
+    ) -> AllocationProgram {
+        let mut prog = AllocationProgram::default();
+
+        prog.import_zstore(zstore);
+        prog.input_expr = vec![(input,)];
+        prog.run();
+
+        println!("{}", prog.relation_sizes_summary());
+
+        assert_eq!(expected_output, prog.output_expr[0].0);
+        prog
+    }
+
+    fn test_aux(input: &str, expected_output: &str) -> AllocationProgram {
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        let expected_output = read_wideptr(&mut zstore, expected_output);
+        test_aux0(&zstore, input, expected_output)
+    }
+
+    fn test_aux1(input: &str, expected_output: WidePtr) -> AllocationProgram {
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        test_aux0(&zstore, input, expected_output)
+    }
+
+    fn test_distilled(input: &str, expected_output: &str, bad_input_output: Option<(&str, &str)>) {
+        // Run the first phase
+        let mut zstore = lurk_zstore();
+        let input = read_wideptr(&mut zstore, input);
+        let expected_output = read_wideptr(&mut zstore, expected_output);
+        let original_program = test_aux0(&zstore, input, expected_output);
+
+        // Start the second phase
+        let mut prog = DistilledAllocationProgram::default();
+
+        // Move the hashes (stored in the allocator) and input over
+        prog.allocator = original_program.allocator.clone();
+        prog.input_expr = original_program.input_expr.clone();
+
+        // transfer over the memory (assume it's been distilled)
+        let raw_memory = original_program.export_memory();
```

Comment:
Eventually, it would be good to introduce a new type for the distilled memory so we don't have to assume.

Reply:
Sorry for the confusing comment, it should've been deleted. There are two types of memory: `VirtualMemory` and `Memory`. The first is the output of the 1st phase, the second is the input to the 2nd phase. 

The behavior of the 1st phase is described by the `LoamProgram` trait and you can implement how to `export_memory`. You have to manually implement `import_memory` for the 2nd phase to be able to import memory. I think the trait organization could be better, but for now you shouldn't be able to accidentally export distilled memory, or import undistilled memory.